### PR TITLE
feat(plugins): Switch to an importlib.metadata.EntryPoint-based plugin loading

### LINKED ISSUE #495: new plugin system
### Description

As discussed on [on issue 287](https://github.com/commitizen-tools/commitizen/issues/287#issuecomment-717540868), we could try [Python Entry Points Explained](https://amir.rachum.com/blog/2017/07/28/python-entry-points/) or [poetry - plugins](https://python-poetry.org/docs/pyproject/#plugins) to implement a new plugin system so that package with `cz_` prefix but not commitzen plugins won't be loaded. (this issue was solved on PR #480 , but refactoring the plugin system might make it easier to maintain in the future)

### Possible Solution

_No response_

### Additional context

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by noirbizarre**
## Description
This PR is proposal implementation for #495

It switches to an `importlib.metadata.EntryPoint`-based plugin loading instead of pattern matching on the module name.
Plugins are now loaded using the `commitizen.plugin` entrypoint while legacy plugins are not loaded anymore but a warning is raised for each legacy plugin seen.

BREAKING CHANGE: Plugins are now exposed as `commitizen.plugin` entrypoints

## Checklist

- [x] Add test cases to all the changes you introduce
- [x] Run `./scripts/format` and `./scripts/test` locally to ensure this change passes linter check and test
- [x] Test the changes on the local machine manually
- [x] Update the documentation for the changes

## Expected behavior

Internal plugins are exposed as `commitizen.plugin` entrypoint. and seen as plugins.
External plugins are seen too as soon as they are exposed as `commitizen.plugin` entrypoint.
Legacy plugins are not loaded and a warning is raised for each legacy plugin detected.

## Steps to Test This Pull Request
 
Try to use any internal plugin without changes in configuration: should work
Try a legacy plugin: KO
Try a plugin migrated to the `commitizen.plugin`: should work

## Additional context

I chose not to have a deprecation warning on this because:
- there was issues with the `cz_` naming
- handling plugins with both `discover_this` and entrypoint expose raise a bit the complexity

But I can switch to this if required:
- register legacy plugins first while raising a deprecation warning
- register entrypoints based plugins, overwritting those registered by legacy pattern
- Add a deprecation warning in the documentation with a target version for support removal

Fixes #495

### CODE REVIEW

**Comment by woile** [comment]
Fantastic PR! Thanks a lot for the contribution.
 
Could you add some documentation? As you can see this would be outdated: https://commitizen-tools.github.io/commitizen/customization/#2-customize-through-customizing-a-class

And if possible a migration guide from the old system, on that same page.

**Comment by noirbizarre** [comment]
There it is, migration section added 👍🏼

**Comment by Lee-W** [comment]
Looks like we'll need to wait for #633 for the CI ?

**Comment by noirbizarre** [comment]
Yes, as soon as #633 is merged, I'll rebase this PR to pass the CI (I also have some other PR not yet submitted)

**Comment by noirbizarre** [comment]
Retargetted and rebased on `v3` branch. CI is passing 🎉 

`importlib_metadata` is used everywhere, even for Python versions having it builtin because:
- each builtin version currently have a different API (ongoing refactoring)
- `MyPy` does not handle typing properly (making tests pass but fail on typing validation)
- this avoid having a `_compat` module introducing behavorial and typing variations between supported Python versions

**Review by Lee-W — APPROVED** [review]
This mechanism looks so great! @woile I plan to merge it later this week. Let me know if you want to take a look.

**Review by woile — APPROVED** [review]
LGTM thanks a lot 💪🏻

**Comment by Lee-W** [comment]
Hi @noirbizarre I'm trying to install the plugins and test it locally. How did you do it on your local side?

**Comment by noirbizarre** [comment]
It depends:
- for `commitizen` built-in plugins, nothing to do except install `commitizen`. PLugins have been mapped to the same names
- for external plugins, they need to be ported as described [here](https://github.com/commitizen-tools/commitizen/blob/v3/docs/customization.md#migrating-from-legacy-plugin-format) (this is quick and easy) and installed (metadata are only visible for installed packages)

I did the migration on https://github.com/noirbizarre/emotional but did not yet push because `v3` is not released (but I will, so you can test)

**Comment by Lee-W** [comment]
> It depends:
> 
>     * for `commitizen` built-in plugins, nothing to do except install `commitizen`. PLugins have been mapped to the same names
> 
>     * for external plugins, they need to be ported as described [here](https://github.com/commitizen-tools/commitizen/blob/v3/docs/customization.md#migrating-from-legacy-plugin-format) (this is quick and easy) and installed (metadata are only visible for installed packages)
> 
> 
> I did the migration on https://github.com/noirbizarre/emotional but did not yet push because `v3` is not released (but I will, so you can test)

What I was trying is to rebase the latest `master` back to `v3` but encounter `commitizen.exceptions.NoCommitizenFoundException: The committer has not been found in the system.` error. Woundring is there anything I did wrong?

**Pull request merged**
